pkg/operator/resource/resourceapply/storage.go

```diff
@@ -103,21 +80,38 @@ func ApplyCSIDriver(ctx context.Context, client storageclientv1.CSIDriversGetter
 		return nil, false, err
 	}
 
-	modified := resourcemerge.BoolPtr(false)
+	metadataModified := resourcemerge.BoolPtr(false)
 	existingCopy := existing.DeepCopy()
+	resourcemerge.EnsureObjectMeta(metadataModified, &existingCopy.ObjectMeta, required.ObjectMeta)
 
-	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, required.ObjectMeta)
-	if !*modified {
-		return existingCopy, false, nil
+	requiredSpecHash := required.Annotations[specHashAnnotation]
+	existingSpecHash := existing.Annotations[specHashAnnotation]
+	sameSpec := requiredSpecHash == existingSpecHash
+	if sameSpec && !*metadataModified {
+		return existing, false, nil
 	}
 
 	if klog.V(4).Enabled() {
 		klog.Infof("CSIDriver %q changes: %v", required.Name, JSONPatchNoError(existing, existingCopy))
 	}
 
-	// TODO: Spec is read-only, so this will fail if user changes it. Should we simply ignore it?
-	actual, err := client.CSIDrivers().Update(ctx, existingCopy, metav1.UpdateOptions{})
-	reportUpdateEvent(recorder, required, err)
+	if sameSpec {
+		// Update metadata by a simple Update call
+		actual, err := client.CSIDrivers().Update(ctx, existingCopy, metav1.UpdateOptions{})
+		reportUpdateEvent(recorder, required, err)
+		return actual, true, err
+	}
+
+	existingCopy.Spec = required.Spec
+	existingCopy.ObjectMeta.ResourceVersion = ""
+	// Spec is read-only after creation. Delete and re-create the object
+	err = client.CSIDrivers().Delete(ctx, existingCopy.Name, metav1.DeleteOptions{})
+	reportDeleteEvent(recorder, existingCopy, err, "Deleting CSIDriver to re-create it with updated parameters")
+	if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
Do we need to poll for `CSIDriver` object to be actually removed from API server before recreating the new CSIDriver object?

Reply:
In case the CSIDriver can't be deleted from any reason (like a finalizer), `Create` below will fail and the operator will retry in the next sync.

I can see that other Apply functions do the same: https://github.com/openshift/library-go/blob/a3f30cb5faf9e47565ff1a6c177592b75aa9f9c1/pkg/operator/resource/resourceapply/core.go#L431-L438


This will result in some errors and the corresponding operator Degraded, if it takes too long time / too many attemtps, but IMO if we poll until a CSIDriver is deleted, it would lead to the same Degraded condition - `ApplyCSIDriver` should report an error if the old CSIDriver exists after a timeout.

Reply:
I updated the code to return something else than a bare `AlreadyExists` from `Create()` below in case there is a finalizer or the deletion did not proceed from any other reason.